vm/interpreter/src/lib.rs

```diff
@@ -49,35 +65,117 @@ where
+                }
+                let ret = self.apply_message(msg)?;
+
+                // Update totals
+                gas_reward += msg.gas_price() * ret.msg_receipt.gas_used;
+                penalty += ret.penalty;
+                receipts.push(ret.msg_receipt);
+
+                // Add callback here if needed in future
+
+                // Add processed Cid to set of processed messages
+                processed.insert(cid);
+                Ok(())
+            };
 
-        for block in &msgs.blocks {
-            // TODO: verifiy ordering of message execution
-
-            for msg in &block.bls_messages {
-                receipts.push(self.apply_message(msg)?);
+            for msg in block.bls_msgs() {
+                process_msg(msg)?;
+            }
+            for msg in block.secp_msgs() {
+                process_msg(msg.message())?;
             }
 
-            for msg in &block.secp_messages {
-                receipts.push(self.apply_message(msg.message())?);
+            // Generate reward transaction for the miner of the block
+            let params = Serialized::serialize(reward::AwardBlockRewardParams {
+                miner: block.header().miner_address().clone(),
+                penalty,
+                gas_reward,
+                // TODO revisit this if/when removed from go clients
+                ticket_count: 1,
+            })?;
+
+            // TODO change this just just one get and update sequence in memory after interop
+            let sys_act = self
+                .state
+                .get_actor(&*SYSTEM_ACTOR_ADDR)?
+                .ok_or_else(|| "Failed to query system actor".to_string())?;
+
+            let rew_msg = UnsignedMessage::builder()
+                .from(SYSTEM_ACTOR_ADDR.clone())
+                .to(REWARD_ACTOR_ADDR.clone())
+                .sequence(sys_act.sequence)
+                .value(BigUint::zero())
+                .gas_price(BigUint::zero())
+                .gas_limit(1 << 30)
+                .method_num(reward::Method::AwardBlockReward as u64)
+                .params(params)
+                .build()?;
+
+            // TODO revisit this ApplyRet structure, doesn't match go logic 1:1 and can be cleaner
+            let ret = self.apply_implicit_message(&rew_msg);
+            if let Some(err) = ret.act_error {
+                return Err(format!(
+                    "failed to apply reward message for miner {}: {}",
+                    block.header().miner_address(),
+                    err
+                )
+                .into());
             }
+
+            // Add callback here for reward message if needed
         }
 
+        // TODO same as above, unnecessary state retrieval
+        let sys_act = self
+            .state
+            .get_actor(&*SYSTEM_ACTOR_ADDR)?
+            .ok_or_else(|| "Failed to query system actor".to_string())?;
+
+        let cron_msg = UnsignedMessage::builder()
+            .to(SYSTEM_ACTOR_ADDR.clone())
```

Comment:
Shouldn't this be to: `CRON_ACTOR_ADDR` from: `SYSTEM_ACTOR_ADDR`?

Reply:
Oops I swapped the order and confused myself lmao